src/Http/WebUtilities/src/HttpResponsePipeWriter.cs

```diff
@@ -0,0 +1,275 @@
+
+        public override Task WriteAsync(string? value)
+        {
+            if (_disposed)
+            {
+                return GetObjectDisposedTask();
+            }
+
+            var length = _encoder.GetByteCount(value, false);
+            var buffer = _writer.GetSpan(length);
+            _encoder.GetBytes(value, buffer, false);
+            _writer.Advance(length);
+
+            return Task.CompletedTask;
+        }
+
+        public override Task WriteAsync(ReadOnlyMemory<char> value, CancellationToken cancellationToken = default)
+        {
+            if (_disposed)
+            {
+                return GetObjectDisposedTask();
+            }
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled(cancellationToken);
+            }
+
+            if (value.IsEmpty)
+            {
+                return Task.CompletedTask;
+            }
+
+            Write(value.Span);
+
+            return Task.CompletedTask;
+        }
+
+        public override Task WriteLineAsync(ReadOnlyMemory<char> value, CancellationToken cancellationToken = default)
+        {
+            if (_disposed)
+            {
+                return GetObjectDisposedTask();
+            }
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled(cancellationToken);
+            }
+
+            if (value.IsEmpty && NewLine.Length == 0)
+            {
+                return Task.CompletedTask;
+            }
+
+            Write(value.Span);
+            Write(NewLine);
+
+            return Task.CompletedTask;
+        }
+
+        public override void Flush()
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(HttpResponseStreamWriter));
+            }
+
+            FlushEncoder();
+            // TOOD: flush
+        }
+
+        // Perf: FlushAsync is invoked to ensure any buffered content is asynchronously written to the underlying
+        // response asynchronously. In its absence, the buffer gets synchronously written to the
+        // response as part of the Dispose which has a perf impact.
+        public override Task FlushAsync()
+        {
+            if (_disposed)
+            {
+                return GetObjectDisposedTask();
+            }
+
+            return FlushInternalAsync().AsTask();
+        }
+
+        public override async ValueTask DisposeAsync()
+        {
+            if (!_disposed)
+            {
+                _disposed = true;
+                await FlushInternalAsync();
+                _writer.Complete();
+            }
+
+            await base.DisposeAsync();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing && !_disposed)
+            {
+                _disposed = true;
+                FlushEncoder();
+                // TOOD: flush
+                _writer.Complete();
```

Comment:
this probably need to be surfaced in the public API (or not done) - since it seems similar to disposing the underlying stream. `HttpResponseStreamWriter` never disposes the stream